:feature: add support for google cloud secret provider.

oryandrive opened the pull request:
Desired Outcome
Introduce support for using Google Cloud Secret Manager as a secret provider.

Implemented Changes
Enhanced get_secret_provider to support selecting the GCP Secret Provider.

Implemented the GCP Secret Provider and a common base class for secret providers.

Added tests and documentation outlining usage and setup instructions.

hub-adda commented on agent_guard_core/credentials/gcp_secrets_manager_provider.py:
```diff
@@ -0,0 +1,167 @@
+from typing import Dict, Optional
+
+from google.cloud import secretmanager
+from google.api_core.exceptions import NotFound, AlreadyExists
```
Could you add this dependency: 'google-api-core' from pypi to the pyproject.toml
pip install google-api-core
and requirements.txt

oryandrive replied:
added  👍🏻

hub-adda commented on agent_guard_core/credentials/gcp_secrets_manager_provider.py:
```diff
@@ -0,0 +1,167 @@
+from google.cloud import secretmanager
+from google.api_core.exceptions import NotFound, AlreadyExists
+
+from .secrets_provider import BaseSecretsProvider, SecretProviderException
```
Could you Run ./code_clean.bash before committing? 
It will run formatting, linting, and security checks
See that the scores are not degrading

oryandrive replied:
Done (: 
`Your code has been rated at 8.73/10 (previous run: 8.72/10, +0.01)`

hub-adda reviewed (commented):
Thanks for the contribution, see my comments

oryandrive commented on agent_guard_core/credentials/gcp_secrets_manager_provider.py:
```diff
@@ -0,0 +1,167 @@
+import json
+from typing import Dict, Optional
```
I provide a sample of how to use it. (:

rafis3 commented:
Hi @oryandrive, thanks a lot for your contribution!
I noticed that integration tests fail when creating the secret, with the following error:

> agent_guard_core.credentials.secrets_provider.SecretProviderException: Error creating secret: Message must be initialized with a dict: google.cloud.secretmanager.v1.CreateSecretRequest

From looking at this [example](https://cloud.google.com/secret-manager/docs/samples/secretmanager-create-secret?hl=en#secretmanager_create_secret-python) I see that the `secret` argument in the secret request should be a dictionary that specifies any of [these](https://cloud.google.com/python/docs/reference/secretmanager/latest/google.cloud.secretmanager_v1.types.Secret) properties.

Could that be a bug?

oryandrive commented:
Hey,
I've addressed all of your comments:
- Ran code clean to ensure the linter is enforcing best practices. 
- Fixed the integration tests and ran them on the GCP dev account to verify everything works as expected.
- Updated the documentation to make it easier for clients to get started.

rafis3 commented on agent_guard_core/credentials/gcp_secrets_manager_provider.py:
```diff
@@ -0,0 +1,145 @@
+
+    def store_secret_dictionary(self, secret_dictionary: Dict[str,
+                                                              str]) -> None:
+        if not secret_dictionary:
```
In the [AWS provider](https://github.com/oryandrive/agent-guard/blob/feature/support-gcp-secret-manager/agent_guard_core/credentials/aws_secrets_manager_provider.py#L93), the check is specifically for `None`, which means it will accept an empty dictionary of `{}`. This code doesn't accept an empty dictionary which makes [this test](https://github.com/oryandrive/agent-guard/blob/feature/support-gcp-secret-manager/tests/integration/gcp_secret_manager_provider_test.py#L23) fail in a clean environment, where the tested GCP secret is empty. Can you please fix the issue?
```suggestion
        # an empty dictionary is valid, yet a none one is not and raises an exception
        if secret_dictionary is None:
```

oryandrive replied:
Thanks, it's fixed now 😊
Let me know if anything else is needed.
I've added documentation and two test cases to ensure both None and empty dictionaries behave as expected.

rafis3 reviewed (changes requested):
@oryandrive added one comment about a bug I found, please take a look.
Also, your original PR had some very nice docstrings for each function, which disappeared after you committed some of your fixes. Would you mind adding them back?

rafis3 commented on tests/integration/gcp_secret_manager_provider_test.py:
```diff
@@ -0,0 +1,142 @@
+
+@pytest.fixture()
+def provider(scope="module"):
+    return GCPSecretsProvider(project_id="mela-50d4f",secret_id="oryantest")
```
Can you please update the initialization of the provider to be more generic?
```suggestion
    return GCPSecretsProvider(project_id="<project ID>")
```

rafis3 reviewed (commented):
@oryandrive last few small comments. Can you please take a final look?

oryandrive commented on agent_guard_core/credentials/gcp_secrets_manager_provider.py:
```diff
@@ -0,0 +1,194 @@
+    def __init__(self,
+                 project_id: str = DEFAULT_PROJECT_ID,
+                 secret_id: str = DEFAULT_SECRET_ID,
+                 secret_id_version: str = DEFAULT_SECRET_VERSION,
```
fixed (:

oryandrive commented:
Thanks, @rafis3  has addressed all your comments! 🙏🏻

rafis3 reviewed (approved):
@oryandrive Looks good, thanks again for your contribution!
